### PR TITLE
Support empty array

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: csrc/type.h [hunk 1]
```diff
@@ -954,7 +962,6 @@ inline PolymorphicValue castToDtype(
   // Cast the given value to the given data type. This enables interface
   // like: IrBuilder::create<Val>(0, DataType::Double) where value is
   // an integer but the desired data type is double.
-  auto value_dtype = getDataType(value);
```
Comment (zasdfgbnm) [inline csrc/type.h:957]: Unused variable

File: csrc/dispatch.cpp [hunk 2]
```diff
@@ -310,7 +312,7 @@ void Expr::dispatch(T handler, Expr* expr) {
     ptr(handler)->handle(expr->as<PipelineCommunication>());
     return;
   }
-  NVF_ERROR(false, "Unknown exprtype in dispatch!");
+  NVF_ERROR(false, "Unknown exprtype in dispatch: ", typeid(*expr).name());
```
Comment (zasdfgbnm) [inline csrc/dispatch.cpp:315]: Unrelated error message improvement

**Comment by zasdfgbnm** [comment]
!build

File: test/test_evaluator.cpp [hunk 3]
```diff
@@ -295,6 +295,19 @@ TEST_F(ExprEvalTest, Array) {
   checkIntValue(evaluator, bb, 5L);
 }
 
+TEST_F(ExprEvalTest, EmptyArray) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto* a = IrBuilder::create<Val>(
```
Comment (liqiangxl) [inline test/test_evaluator.cpp:302]: If a wrong array length is passed as input, e.g.
```
  auto* a = IrBuilder::create<Val>(
      std::vector<int64_t>{},
      ArrayType{std::make_shared<DataType>(DataType::Int), 2});
```
the err msg still shows `Empty array is not supported`.

File: csrc/type.h [hunk 4]
```diff
@@ -488,6 +488,14 @@ inline bool hasCompatibleDataType(
     }
     auto ptr = std::get<PointerType>(dtype.type);
     return dataTypeSize(*ptr.type) == value.as<Pointer>().size();
+  } else if (std::holds_alternative<ArrayType>(dtype.type)) {
+    if (!value.is<std::vector>()) {
+      return false;
+    }
+    if (std::get<ArrayType>(dtype.type).size == 0 &&
```
Comment (naoyam) [inline csrc/type.h:495]: Why this is considered compatible? IIUC, we know both `dtype` and `value` are arrays, but we haven't checked the element type of each array. However, if both are empty, it seems we consider they are compatible. Why?

Comment (zasdfgbnm) [reply csrc/type.h:495]: For polymorphic value, an empty list has no item type. Just like for python, `[]` is just an empty list, not an "empty list of integers" or "empty list of strings"

File: test/test_evaluator.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply test/test_evaluator.cpp:311]: Fixed. I added a check about if the size matches.

File: csrc/type.h [hunk 5]
```diff
@@ -488,6 +488,14 @@ inline bool hasCompatibleDataType(
...
+        value.as<std::vector>().empty()) {
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/type.h]: Fixed: https://github.com/NVIDIA/Fuser/pull/837#discussion_r1316312273

**Review by liqiangxl — APPROVED** [review]
LGTM.

**Pull request merged**